corehq/sql_db/shard_data_management.py

```diff
@@ -0,0 +1,44 @@
+from __future__ import absolute_import
+from django.db import connections
+from corehq.sql_db.config import partition_config
+
+
+def get_count_of_unmatched_models_by_shard(database, model):
+    """
+    Get counts of any `model` data residing on shards that aren't expected to live in `database`.
+
+    Returns a list of tuples of the format:
+    [
+      (shard_id, count_of_unexpected_models),
+    ]
+
+    The list will be empty if no invalid data is found.
+    """
+    cursor = connections[database].cursor()
+    query = _get_shard_count_query(model)
+    valid_shards = partition_config.get_shards_on_db(database)
+    cursor.execute(query, [valid_shards])
+    results = cursor.fetchall()
+    return results
+
+
+def _get_shard_count_query(model):
+    # have to cast to varchar because some tables have uuid types
+    shard_id_function = "hash_string({id_field}::varchar, 'siphash24') & {total_shard_count}".format(
```

Comment:
did the casting to varchar work as expected for the uuid types? i remember having to cast to byte array: https://github.com/dimagi/commcare-hq/blob/b9c453bf26d6b0267b0e801ad9a3ba9c61119a39/corehq/form_processor/backends/sql/dbaccessors.py#L104

Reply:
it ran but am not sure it worked properly because I don't have any local messaging data. 

had been debating whether it was worth the overhead of getting a test setup up and running for these functions. it's a lot of bootstrapping code but is probably worth it if we're going to be relying on them to validate/delete data.

Reply:
yeah we'll want to make sure we test for this use case if we're going to rely on it for validating or deleting data.

casting the uuid to varchar is not going to yield the right answer - we'll need to find a way to cast the uuid type to bytea to match what this function is doing: https://github.com/dimagi/commcare-hq/blob/a40ce735e7dd8f9d96bc69b273e44035cde928b1/corehq/form_processor/backends/sql/dbaccessors.py#L117-L128

Reply:
thanks for flagging this. probably saved me loads of future time trying to figure out what was going wrong down the line: https://github.com/dimagi/commcare-hq/pull/18409/commits/5684b9bada9070e6e1d01eb503306c9c9b63fbc5